### DISCUSSION TITLE
gatherer sbd_config with argument 'SBD_DEVICE'

### DISCUSSION THREAD

**Discussion by angelabriel** [trento-project/wanda#117] [Q&A] [answered]
If I understand the docu correctly it returns a list of the 'SBD_DEVICE' entry found in the config file. (/dev/vdc;/dev/vdb)
Is this an 'array' or a 'string'?

If it is an array, how can this be used in the 'expectation' property, if I need the length of the array (which means I need the number of returned devices)?
Can I use 'len()' directly for the 'facts' resolved values like
       expect: facts.sbd_multiple_SBD_DEVICE.len() == values.expected_multiple_SBD_DEVICE

or do I need to assign 'facts.sbd_multiple_SBD_DEVICE' to a variable (where in the yaml file?) and use this variable for the expression
       let sbddev_list = facts.sbd_multiple_SBD_DEVICE
       expect: sbddev_list.len() == values.expected_multiple_SBD_DEVICE

It's not really clear to me, where (and how) I can use the possibilities of the expression language inside the  'expectation' property.

**Comment by pirat013** [comment]
I've a similar problem with the check corosync has at least two rings configured. If I could get advice on how such a complex check could be solved. Any help is appreciated.

**Reply by fabriziosestito** [reply]
@pirat013 could you post an example of what you are trying to achieve?

**Reply by pirat013** [reply]
This is the section in the corosync.conf:
`
nodelist {
    node {
            ring0_addr: 192.168.1.11
            ring1_addr: 192.168.2.11
            nodeid: 1
    }
    node {
            ring0_addr: 192.168.1.12
            ring1_addr: 192.168.2.12
            nodeid: 2
    }`
The information that two rings are configured and the parameter seams valid must be extracted with the gatherer. 
The result should show the corosync.conf file has two rings configured which might work if the corosync is starting.

**Reply by fabriziosestito** [reply]
@pirat013 sorry, I am afraid I am not understanding the context. Could you elaborate on this a bit more with facts, values, and expectations examples also in pseudocode in a new discussion?

**Reply by pirat013** [reply]
@fabriziosestito thanks I got this. I'll open a new discussion if needed.

**Comment by fabriziosestito** [comment, marked as answer]
@angelabriel the SBD_DEVICE value is a semicolon separated string, so we can leverage the evaluation language by doing something like: 
```rust
facts.sbd_devices.split(";").len()
```

where `sbd_devices` is the fact containing `sbd_config` gathered fact with argument `SBD_DEVICE`

**Reply by angelabriel** [reply]
Sounds good, thanks.
Any estimation when such an enhancement of the evaluation language will be available for our check rewrites?

**Reply by fabriziosestito** [reply]
@angelabriel this is already implemented

**Reply by angelabriel** [reply]
Thanks a lot, works perfectly as needed.
